src/iop/sigmoid.c

```diff
@@ -0,0 +1,631 @@
+  }
+}
+
+void process_loglogistic_per_channel_interpolated(dt_dev_pixelpipe_iop_t *piece, const void *const ivoid, void *const ovoid,
+                             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  const dt_iop_sigmoid_data_t *module_data = (dt_iop_sigmoid_data_t *)piece->data;
+
+  const float *const in = (const float *)ivoid;
+  float *const out = (float *)ovoid;
+  const size_t npixels = (size_t)roi_in->width * roi_in->height;
+
+  const float white_target = module_data->white_target;
+  const float paper_exp = module_data->paper_exposure;
+  const float film_fog = module_data->film_fog;
+  const float contrast_power = module_data->film_power;
+  const float skew_power = module_data->paper_power;
+  const float hue_preservation = module_data->hue_preservation;
+
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(npixels, white_target, paper_exp, film_fog, contrast_power, skew_power, hue_preservation) \
+  dt_omp_sharedconst(in, out) \
+  schedule(static)
+#endif
+  for(size_t k = 0; k < 4 * npixels; k += 4)
+  {
+    const float *const restrict pix_in = in + k;
+    float *const restrict pix_out = out + k;
+    float DT_ALIGNED_ARRAY pix_in_strict_positive[4];
+    float DT_ALIGNED_ARRAY per_channel[4];
+
+    // Force negative values to zero
+    desaturate_negative_values(pix_in, pix_in_strict_positive);
+
+    for(size_t c = 0; c < 3; c++)
+    {
+      per_channel[c] = generalized_loglogistic_sigmoid(pix_in_strict_positive[c], white_target, paper_exp, film_fog, contrast_power, skew_power);
+    }
+
+    // Hue correction by scaling the middle value relative to the max and min values.
+    dt_iop_sigmoid_value_order_t pixel_value_order;
+    pixel_channel_order(pix_in_strict_positive, &pixel_value_order);
+    preserve_hue_and_energy(pix_in_strict_positive, per_channel, pix_out, pixel_value_order, hue_preservation);
+
+    // Copy over the alpha channel
+    pix_out[3] = pix_in[3];
+  }
+}
+
+
+
+/** process, all real work is done here. */
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const ivoid, void *const ovoid,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  // this is called for preview and full pipe separately, each with its own pixelpipe piece.
+  dt_iop_sigmoid_data_t *module_data = (dt_iop_sigmoid_data_t *)piece->data;
+
+  if (module_data->color_processing == DT_SIGMOID_METHOD_PER_CHANNEL)
+  {
+    if (module_data->hue_preservation >= 0.001f)
+    {
+      process_loglogistic_per_channel_interpolated(piece, ivoid, ovoid, roi_in, roi_out);
+    }
+    else
+    {
+      process_loglogistic_per_channel(piece, ivoid, ovoid, roi_in, roi_out);
+    }
+  }
+  else
+  {
+    process_loglogistic_rgb_ratio(piece, ivoid, ovoid, roi_in, roi_out);
+  }
+}
+
+void init_pipe(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
```

Comment:
can be removed as default implementation

Reply:
I get
```
free(): invalid next size (fast)
Aborted (core dumped)
```
If this and the next function is removed.

Reply:
Strange, the default implementations are:

```
static void default_init_pipe(struct dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe,
                              dt_dev_pixelpipe_iop_t *piece)
{
  piece->data = calloc(1,self->params_size);
}

static void default_cleanup_pipe(struct dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe,
                                 dt_dev_pixelpipe_iop_t *piece)
{
  free(piece->data);
}

```

Reply:
My bad, dt_iop_sigmoid_data_t is not equivalent in size to dt_iop_sigmoid_params_t !!!

So you need to keep `init_pipe()`. You still can remove `cleanup_pipe()`.